langgraph/prebuilt/chat_agent_executor.py

```diff
@@ -0,0 +1,128 @@
+import json
+import operator
+from typing import Annotated, Sequence, TypedDict
+
+from langchain.tools.render import format_tool_to_openai_function
+from langchain_core.agents import AgentAction
+from langchain_core.messages import BaseMessage, FunctionMessage
+from langchain_core.runnables import RunnableLambda
+
+from langgraph.graph import END, StateGraph
+from langgraph.prebuilt.tool_executor import ToolExecutor
+
+
+def create_function_calling_executor(model, tools):
+    if isinstance(tools, ToolExecutor):
+        tool_executor = tools
+        tool_classes = tools.tools
+    else:
+        tool_executor = ToolExecutor(tools)
+        tool_classes = tools
+    model = model.bind_functions(
+        [format_tool_to_openai_function(t) for t in tool_classes]
+    )
+
+    # Define the function that determines whether to continue or not
+    def should_continue(state):
+        messages = state["messages"]
+        last_message = messages[-1]
+        # If there is no function call, then we finish
+        if "function_call" not in last_message.additional_kwargs:
```

Comment:
note in passing: we should figure out how to expose an interface around this... `additional_kwargs` feels like should be hidden from end-user

Reply:
yup agree